File: src/neetcode/Java/_20ValidParentheses.java
package neetcode.Java;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Stack;

public class _20ValidParentheses {
    public boolean isValid(String s) {
        if (s.length() % 2 == 1) return false;
        if (s.charAt(0) == ')' || s.charAt(0) == '}' || s.charAt(0) == ']') return false;
        HashMap<Character, Character> map = new HashMap<>();
        map.put(']', '[');
        map.put('}', '{');
        map.put(')', '(');
        Stack<Character> st = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i))) {
                if (!st.isEmpty() && map.get(s.charAt(i)) == st.peek()) st.pop();
                else
                    return false;
            } else st.push(s.charAt(i));
        }
        return st.isEmpty();
    }


    public static void main(String[] args) {
        _20ValidParentheses obj = new _20ValidParentheses();
        System.out.println(obj.isValid("(){}}{"));
    }
}

File: src/neetcode/kotlin/_20ValidParentheses.kt
package neetcode.kotlin

import java.util.*

class _20ValidParentheses {
    fun isValid(s: String): Boolean {
        if (s.length % 2 == 1) return false
        if (s[0] == ')' || s[0] == '}' || s[0] == ']') return false
        val map = HashMap<Char, Char>()
        map[']'] = '['
        map['}'] = '{'
        map[')'] = '('
        val st = Stack<Char>()
        for (i in 0 until s.length) {
            if (map.containsKey(s[i])) {
                if (!st.isEmpty() && map[s[i]] === st.peek()) st.pop()
                else return false
            } else st.push(s[i])
        }
        return st.isEmpty()
    }


    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val obj = _20ValidParentheses()
            println(obj.isValid("(){}}{"))
        }
    }
}
